sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -17,17 +17,151 @@
+
+import org.apache.spark.sql.{AnalysisException, SaveMode}
+import org.apache.spark.sql.catalyst.TableIdentifier
 import org.apache.spark.sql.catalyst.analysis.MultiInstanceRelation
-import org.apache.spark.sql.catalyst.expressions.AttributeReference
-import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, Statistics}
-import org.apache.spark.sql.sources.v2.reader._
+import org.apache.spark.sql.catalyst.expressions.{AttributeReference, Expression}
+import org.apache.spark.sql.catalyst.plans.logical.{LeafNode, LogicalPlan, Statistics}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.sources.{DataSourceRegister, Filter}
+import org.apache.spark.sql.sources.v2.{DataSourceOptions, DataSourceV2, ReadSupport, ReadSupportWithSchema, WriteSupport}
+import org.apache.spark.sql.sources.v2.reader.{DataSourceReader, SupportsPushDownCatalystFilters, SupportsPushDownFilters, SupportsPushDownRequiredColumns, SupportsReportStatistics}
+import org.apache.spark.sql.sources.v2.writer.DataSourceWriter
+import org.apache.spark.sql.types.StructType
 
 case class DataSourceV2Relation(
-    fullOutput: Seq[AttributeReference],
-    reader: DataSourceReader)
-  extends LeafNode with MultiInstanceRelation with DataSourceReaderHolder {
+    source: DataSourceV2,
+    options: Map[String, String],
+    path: Option[String] = None,
+    table: Option[TableIdentifier] = None,
+    projection: Option[Seq[AttributeReference]] = None,
+    filters: Option[Seq[Expression]] = None,
+    userSchema: Option[StructType] = None) extends LeafNode with MultiInstanceRelation {
+
+  override def simpleString: String = {
+    "DataSourceV2Relation(" +
+      s"source=$sourceName${path.orElse(table).map(loc => s"($loc)").getOrElse("")}, " +
+      s"schema=[${output.map(a => s"$a ${a.dataType.simpleString}").mkString(", ")}], " +
+      s"filters=[${pushedFilters.mkString(", ")}] options=$options)"
+  }
+
+  override lazy val schema: StructType = reader.readSchema()
+
+  override lazy val output: Seq[AttributeReference] = {
+    projection match {
+      case Some(attrs) =>
+        // use the projection attributes to avoid assigning new ids. fields that are not projected
+        // will be assigned new ids, which is okay because they are not projected.
+        val attrMap = attrs.map(a => a.name -> a).toMap
+        schema.map(f => attrMap.getOrElse(f.name,
+          AttributeReference(f.name, f.dataType, f.nullable, f.metadata)()))
+      case _ =>
+        schema.toAttributes
+    }
+  }
+
+  private lazy val v2Options: DataSourceOptions = {
+    // ensure path and table options are set correctly
+    val updatedOptions = new mutable.HashMap[String, String]
+    updatedOptions ++= options
+
+    path match {
+      case Some(p) =>
+        updatedOptions.put("path", p)
+      case None =>
+        updatedOptions.remove("path")
+    }
+
+    table.map { ident =>
+      updatedOptions.put("table", ident.table)
```

Comment:
this is something I wanna avoid. At the end we still need to define some standard option keys, like `path` and `table`. It would be good if we can provide some methods for end users and data source developers to easily write/read these standard options. I agree this can be done in a different PR.

Reply:
I think we agree here. I want to avoid doing this outside of either `DataSourceOptions` or `DataSourceV2Relation`. If we can create `DataSourceOptions` from a `Option[TableIdentifier]` and add the `getTable` accessor, then that works for me. My main motivation is to avoid having this piece of code copied throughout the SQL planner.

Reply:
Opened [SPARK-23341](https://issues.apache.org/jira/browse/SPARK-23341) for this.